internal/venice-client-common/src/main/java/com/linkedin/venice/stats/MetricDataRangeDecorator.java

```diff
@@ -0,0 +1,302 @@
+            return createPointCollection((GaugeData<LongPointData>) originalData);
+          }
+        };
+
+      case DOUBLE_SUM:
+        return new SumData<DoublePointData>() {
+          @Override
+          public AggregationTemporality getAggregationTemporality() {
+            return ((ImmutableSumData<DoublePointData>) originalData).getAggregationTemporality();
+          }
+
+          @Override
+          public Collection<DoublePointData> getPoints() {
+            return createPointCollection((ImmutableSumData<DoublePointData>) originalData);
+          }
+
+          @Override
+          public boolean isMonotonic() {
+            return ((ImmutableSumData<DoublePointData>) originalData).isMonotonic();
+          }
+        };
+
+      case LONG_SUM:
+        return new SumData<LongPointData>() {
+          @Override
+          public AggregationTemporality getAggregationTemporality() {
+            return ((ImmutableSumData<LongPointData>) originalData).getAggregationTemporality();
+          }
+
+          @Override
+          public Collection<LongPointData> getPoints() {
+            return createPointCollection((ImmutableSumData<LongPointData>) originalData);
+          }
+
+          @Override
+          public boolean isMonotonic() {
+            return ((ImmutableSumData<LongPointData>) originalData).isMonotonic();
+          }
+        };
+
+      case HISTOGRAM:
+        return new HistogramData() {
+          @Override
+          public AggregationTemporality getAggregationTemporality() {
+            return ((HistogramData) originalData).getAggregationTemporality();
+          }
+
+          @Override
+          public Collection<HistogramPointData> getPoints() {
+            return createPointCollection((HistogramData) originalData);
+          }
+        };
+
+      case SUMMARY:
+        return new SummaryData() {
+          @Override
+          public Collection<SummaryPointData> getPoints() {
+            return createPointCollection((SummaryData) originalData);
+          }
+        };
+
+      default:
+        throw new IllegalStateException("Unsupported MetricData type: " + originalMetricData.getType());
+    }
+  }
+
+  /**
+   * Override the open source method only to help type cast to SumData.
+   */
+  @Override
+  public SumData<DoublePointData> getDoubleSumData() {
+    if (getType() == MetricDataType.DOUBLE_SUM) {
+      return (SumData<DoublePointData>) getData();
+    }
+    return ImmutableSumData.empty();
+  }
+
+  /**
+   * Creates a collection that wraps the original points and applies the range filter.
+   */
+  private <T extends PointData> Collection<T> createPointCollection(Data<T> originalData) {
+    return new AbstractCollection<T>() {
+      @Override
+      public Iterator<T> iterator() {
+        return new RangeIterator<>(originalData.getPoints().iterator(), startIndex, endIndex);
+      }
+
+      @Override
+      public int size() {
+        return Math.max(0, endIndex - startIndex);
```

Comment:
normally, size should be `endIndex - startIndex + 1` for a 0-indexing array. Does it have different meaning here?